collector/zfs.go

```diff
@@ -0,0 +1,114 @@
+package collector
+
+// +build linux freebsd
+// +build !nozfs
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+type zfsMetricValue int
+
+const zfsErrorValue = zfsMetricValue(-1)
+
+var zfsNotAvailableError = errors.New("ZFS / ZFS statistics are not available")
+
+type zfsSysctl string
+type zfsSubsystemName string
+
+const (
+	arc            = zfsSubsystemName("zfsArc")
+	zpoolSubsystem = zfsSubsystemName("zfsPool")
+)
+
+// Metrics
+
+type zfsMetric struct {
+	subsystem zfsSubsystemName // The Prometheus subsystem name.
+	name      string           // The Prometheus name of the metric.
+	sysctl    zfsSysctl        // The sysctl of the ZFS metric.
+}
+
+type datasetMetric struct {
+	subsystem zfsSubsystemName
+	name      string
+}
+
+// Collector
+
+func init() {
+	Factories["zfs"] = NewZFSCollector
+}
+
+type zfsCollector struct {
+	zfsMetrics []zfsMetric
+}
+
+func NewZFSCollector() (Collector, error) {
+	return &zfsCollector{}, nil
+}
+
+func (c *zfsCollector) Update(ch chan<- prometheus.Metric) (err error) {
+
```

Comment:
I've noticed you sometimes add extra whitespace at the beginning or end of function blocks.  To me, it looks a little bit messy.  However, I'm not a member of the Prometheus project, so I will defer to their stylistic preferences.